### ISSUE TITLE
Envoy egress bpf_metadata filter looking up conntrack global map with 127.0.0.1

### ISSUE THREAD

**Issue by fangpenlin** [cilium/cilium#29928]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Could be a similar bug to this one #29406, but different error message printed. When I setup a dedicated ingress with NodePort as the service type, and try to send HTTP request to that node port, I saw a 503 error after the connection hangs for a few seconds:

```
curl -v http://localhost:30109/ -v
*   Trying [::1]:30109...
* connect to ::1 port 30109 failed: Connection refused
*   Trying 127.0.0.1:30109...
* Connected to localhost (127.0.0.1) port 30109
> GET / HTTP/1.1
> Host: localhost:30109
> User-Agent: curl/8.3.0
> Accept: */*
>
< HTTP/1.1 503 Service Unavailable
< content-length: 91
< content-type: text/plain
< date: Fri, 15 Dec 2023 20:35:46 GMT
< server: envoy
<
* Connection #0 to host localhost left intact
upstream connect error or disconnect/reset before headers. reset reason: connection failure
```

then I can see Envoy proxy prints out error message like this immediately:

```
[info][filter] [cilium/conntrack.cc:229] cilium.bpf_metadata: IPv4 conntrack map global lookup failed: No such file or directory
```

After looking into it deeper with debug log level, I realized that somehow Cilium envoy filter is looking up record in conntrack global map with `127.0.0.1` as the source and dest IP addresses:

```
[2023-12-15 20:37:52.946][17][debug][misc] [cilium/bpf_metadata.cc:215] EGRESS POD IP: 127.0.0.1, destination IP: 127.0.0.1
[2023-12-15 20:37:52.946][17][debug][filter] [cilium/conntrack.cc:178] cilium.bpf_metadata: Using conntrack map global
[2023-12-15 20:37:52.946][17][trace][filter] [cilium/conntrack.cc:197] cilium.bpf_metadata: Looking up key: 7f000001, 7f000001, 288d, d51c, 6, 0
[2023-12-15 20:37:52.946][17][debug][filter] [cilium/conntrack.cc:128] cilium.bpf_metadata: Opened IPv4 conntrack map global
[2023-12-15 20:37:52.946][17][info][filter] [cilium/conntrack.cc:229] cilium.bpf_metadata: IPv4 conntrack map global lookup failed: No such file or directory
[2023-12-15 20:37:52.946][17][trace][filter] [cilium/ipcache.cc:82] cilium.ipcache: Looking up key: 7f000001, prefixlen: 32
[2023-12-15 20:37:52.946][17][debug][filter] [cilium/ipcache.cc:90] cilium.ipcache: 127.0.0.1 has ID 2
[2023-12-15 20:37:52.946][17][trace][filter] [cilium/ipcache.cc:82] cilium.ipcache: Looking up key: 7f000001, prefixlen: 32
[2023-12-15 20:37:52.946][17][debug][filter] [cilium/ipcache.cc:90] cilium.ipcache: 127.0.0.1 has ID 2
[2023-12-15 20:37:52.946][17][trace][filter] [cilium/ipcache.cc:82] cilium.ipcache: Looking up key: a1401c6, prefixlen: 32
[2023-12-15 20:37:52.946][17][debug][filter] [cilium/ipcache.cc:90] cilium.ipcache: 10.20.1.198 has ID 8
[2023-12-15 20:37:52.946][17][debug][filter] [./cilium/socket_option.h:184] Cilium SocketOption(): source_identity: 8, ingress: false, port: 10381, pod_ip: 10.20.1.198, source_addresses: /10.20.1.198:0/, mark: 80b00 (magic mark: b00, cluster: 0, ID: 8)
```

I think it doesn't make too much sense as I anticipate the conntrack look up should be using the actual curl client's source IP and the dest pod's IP.

### Cilium Version

1.14.5

### Kernel Version

Linux um790-00 6.1.55 #1-NixOS SMP PREEMPT_DYNAMIC Sat Sep 23 09:11:13 UTC 2023 x86_64 GNU/Linux

### Kubernetes Version

Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.2

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by judge-red** [comment]
Hm, I also see the `cilium.bpf_metadata: IPv4 conntrack map global lookup failed: No such file or directory` error, but the rest looks different for me. Not sure I'm trying to do the same thing, though. I'm very new to Cilium and not sure how to debug my issue, so maybe I'm looking at a config error on my part - but I found this issue when googling for the error above :)

```
$ curl <URL of ingress> -I
HTTP/1.1 400 Bad Request
content-length: 11
content-type: text/plain
date: Wed, 03 Jan 2024 10:28:42 GMT
server: envoy
connection: close
```

```
cilium-envoy-dxx4r cilium-envoy [2024-01-03 10:23:40.613][17][debug][misc] [cilium/bpf_metadata.cc:215] EGRESS POD IP: 192.168.0.182, destination IP: 192.168.0.218
cilium-envoy-dxx4r cilium-envoy [2024-01-03 10:23:40.613][17][debug][filter] [cilium/conntrack.cc:178] cilium.bpf_metadata: Using conntrack map global
cilium-envoy-dxx4r cilium-envoy [2024-01-03 10:23:40.613][17][trace][filter] [cilium/conntrack.cc:197] cilium.bpf_metadata: Looking up key: c0a800da, c0a800b6, 77a6, f3bd, 6, 0
cilium-envoy-dxx4r cilium-envoy [2024-01-03 10:23:40.616][17][debug][filter] [cilium/conntrack.cc:128] cilium.bpf_metadata: Opened IPv4 conntrack map global
cilium-envoy-dxx4r cilium-envoy [2024-01-03 10:23:40.616][17][info][filter] [cilium/conntrack.cc:229] cilium.bpf_metadata: IPv4 conntrack map global lookup failed: No such file or directory
```

**Comment by judge-red** [comment]
nvm, my issue was actually that I had proxy-protocol enabled on the LoadBalancer Service but not on the envoy - well, support for enabling proxy protocol on the cilium-envoy was only added post-1.14

**Comment by squeed** [comment]
@fangpenlin thanks for the report. Could you supply your ingress configuration exactly?

Separately, making requests to `localhost` can be fragile. What happens if you make a request to the NodePort externally?

**Comment by jschwartzy** [comment]
We're also encountering a similar issue with Envoy via the `GatewayAPI`.
We can narrow it down to when Envoy is on the name node as the destination pod, we hit these `503` errors. 

To reproduce this issue:

- Deploy an HTTP application on K8s with a standard Service definition. To simplify repro, just have a single replica.
- Deploy Cilium with Envoy either embedded or daemonset
- SSH onto the node where the application is running and try to curl the `NodePort`
- Or you can Curl the NodeIP with `NodePort` (external)
- Or you can Curl the LoadBalancer configured to forward to the `NodePort`. Depending on the Cloud infra in question, it helps to narrow/reduce the targets to the node where the app replica is running. (For example, I've labeled the node with `cilium-debug=true` and then set the annotation to `service.beta.kubernetes.io/aws-load-balancer-target-node-labels: cilium-debug=true`).
- Any of those 3 scenarios results in the same output (example below, curling LoadBalancer)

```
curl -v http://<redacted>.com/ip
*   Trying 10.x.x.x:80...
* Connected to <redacted>.com (10.x.x.x) port 80 (#0)
> GET /ip HTTP/1.1
> Host: <redacted>.com
> User-Agent: curl/7.79.1
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 503 Service Unavailable
< content-length: 91
< content-type: text/plain
< date: Tue, 30 Jan 2024 16:58:44 GMT
< server: envoy
< 
* Connection #0 to host <redacted>.com left intact
upstream connect error or disconnect/reset before headers. reset reason: connection failure
```

Snippet from envoy debug log:
```
[2024-01-30 16:56:23.786][12][trace][filter] [external/envoy/source/extensions/filters/listener/tls_inspector/tls_inspector.cc:86] tls inspector: new connection accepted
[2024-01-30 16:56:23.787][12][trace][filter] [external/envoy/source/common/network/listener_filter_buffer_impl.cc:95] onFileEvent: 5
[2024-01-30 16:56:23.787][12][trace][main] [external/envoy/source/common/event/dispatcher_impl.cc:250] item added to deferred deletion list (size=1)
[2024-01-30 16:56:23.787][12][trace][main] [external/envoy/source/common/event/dispatcher_impl.cc:125] clearing deferred deletion list (size=1)
[2024-01-30 16:56:23.938][13][debug][pool] [external/envoy/source/common/conn_pool/conn_pool_base.cc:793] [C280] connect timeout
[2024-01-30 16:56:23.938][13][debug][connection] [external/envoy/source/common/network/connection_impl.cc:139] [C280] closing data_to_write=0 type=1
[2024-01-30 16:56:23.938][13][debug][connection] [external/envoy/source/common/network/connection_impl.cc:250] [C280] closing socket: 1
[2024-01-30 16:56:23.939][13][trace][connection] [external/envoy/source/common/network/connection_impl.cc:423] [C280] raising connection event 1
[2024-01-30 16:56:23.939][13][debug][client] [external/envoy/source/common/http/codec_client.cc:107] [C280] disconnect. resetting 0 pending requests
[2024-01-30 16:56:23.939][13][debug][pool] [external/envoy/source/common/conn_pool/conn_pool_base.cc:484] [C280] client disconnected, failure reason: 
[2024-01-30 16:56:23.939][13][debug][router] [external/envoy/source/common/router/router.cc:1290] [C279][S16061982579707376886] upstream reset: reset reason: connection failure, transport failure reason: 
[2024-01-30 16:56:23.939][13][trace][main] [external/envoy/source/common/event/dispatcher_impl.cc:250] item added to deferred deletion list (size=1)
[2024-01-30 16:56:23.939][13][trace][main] [external/envoy/source/common/event/dispatcher_impl.cc:250] item added to deferred deletion list (size=2)
[2024-01-30 16:56:23.939][13][debug][http] [external/envoy/source/common/http/filter_manager.cc:996] [C279][S16061982579707376886] Sending local reply with details upstream_reset_before_response_started{connection_failure}
[2024-01-30 16:56:23.939][13][trace][misc] [external/envoy/source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x55ebf7ff340 for 300000ms, min is 300000ms
[2024-01-30 16:56:23.939][13][trace][http] [external/envoy/source/common/http/filter_manager.cc:1158] [C279][S16061982579707376886] encode headers called: filter=cilium.l7policy status=0
[2024-01-30 16:56:23.940][13][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:1773] [C279][S16061982579707376886] encoding headers via codec (end_stream=false):
':status', '503'
'content-length', '91'
'content-type', 'text/plain'
'date', 'Tue, 30 Jan 2024 16:56:23 GMT'
'server', 'envoy'
```

Note: the routing seems to work fine when the Envoy pod is on a different node from the app replica.

**Comment by jschwartzy** [comment]
Some additional observations, providing Hubble traces:

1.  Envoy proxy is on a different node than the app replica. Note that all traffic is directed to the pod replica
```
Feb  3 06:18:07.531: 100.x.x.x:38579 (ingress) -> test/httpbin-alb-7c54bb4487-hhch2:8080 (ID:2611) to-proxy FORWARDED (TCP Flags: SYN)
Feb  3 06:18:07.532: 100.x.x.x:38579 (ingress) -> test/httpbin-alb-7c54bb4487-hhch2:8080 (ID:2611) to-proxy FORWARDED (TCP Flags: ACK)
Feb  3 06:18:07.532: 10.x.x.x:53011 (ingress) -> test/httpbin-alb-7c54bb4487-hhch2:8080 (ID:2611) http-request FORWARDED (HTTP/1.1 GET http://10.x.x.x:31666/)
Feb  3 06:18:07.532: 100.x.x.x:38579 (ingress) -> test/httpbin-alb-7c54bb4487-hhch2:8080 (ID:2611) to-proxy FORWARDED (TCP Flags: ACK, PSH)
Feb  3 06:18:07.539: 10.x.x.x:45644 (ingress) -> test/httpbin-alb-7c54bb4487-hhch2:8080 (ID:2611) to-endpoint FORWARDED (TCP Flags: SYN)
Feb  3 06:18:07.539: 10.x.x.x:45644 (ingress) -> test/httpbin-alb-7c54bb4487-hhch2:8080 (ID:2611) to-endpoint FORWARDED (TCP Flags: ACK)
Feb  3 06:18:07.542: 100.x.x.x:38579 (ingress) -> test/httpbin-alb-7c54bb4487-hhch2:8080 (ID:2611) http-request FORWARDED (HTTP/1.1 GET http://10.x.x.x:31666/)
Feb  3 06:18:07.545: 10.x.x.x:45644 (ingress) -> test/httpbin-alb-7c54bb4487-hhch2:8080 (ID:2611) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Feb  3 06:18:07.553: 100.x.x.x:38579 (ingress) <- test/httpbin-alb-7c54bb4487-hhch2:8080 (ID:2611) http-response FORWARDED (HTTP/1.1 200 17ms (GET http://10.x.x.x:31666/))
Feb  3 06:18:07.557: 10.x.x.x:53011 (ingress) <- test/httpbin-alb-7c54bb4487-hhch2:8080 (ID:2611) http-response FORWARDED (HTTP/1.1 200 26ms (GET http://10.x.x.x:31666/))
Feb  3 06:18:09.550: 10.x.x.x:45644 (ingress) -> test/httpbin-alb-7c54bb4487-hhch2:8080 (ID:2611) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
```
2. Envoy proxy on the same node as the app replica. Note that it tries to use the LoadBalancer service, not the underlying pod. Incidentally, if one tries to curl the LoadBalancer service from inside the cluster, the request times out.
```
Feb  3 06:16:24.808: 100.x.x.x:37889 (ingress) -> test/httpbin-alb-7c54bb4487-hhch2:8080 (ID:2611) to-proxy FORWARDED (TCP Flags: SYN)
Feb  3 06:16:29.802: 10.x.x.x:53007 (ingress) -> test/cilium-gateway-gateway-nlb-test:31666 (world) http-request FORWARDED (HTTP/1.1 GET http://10.x.x.x:31666/)
Feb  3 06:16:29.803: 10.x.x.x:53007 (ingress) <- test/cilium-gateway-gateway-nlb-test:31666 (world) http-response FORWARDED (HTTP/1.1 503 4997ms (GET http://10.x.x.x:31666/))
```
Service:
```
> kubectl get svc cilium-gateway-gateway-nlb-test 
NAME                              TYPE           CLUSTER-IP       EXTERNAL-IP                             PORT(S)                      AGE
cilium-gateway-gateway-nlb-test   LoadBalancer   172.x.x.x   k8s-test-ciliumga-<redacted>.amazonaws.com   80:31666/TCP                 46h
```

**Comment by acelinkio** [comment]
After hours of research into tickets related to `upstream connect error or disconnect/reset before headers. reset reason: connection timeout` errors.  This issue appears to be the closest to the cause.  I can confirm @jschwartzy's findings.  There is always one Cilium daemonset pod that is unable to deliver gateway traffic to workloads on the same node.  

I have also worked directly with some other community members to enable gateway on their cluster and they experienced the same issue.

Example where 'bad' cilium pod cannot deliver to workload on the same node.
```log
hubble observe -f --identity 8
Feb  5 21:17:44.348: 10.42.1.45:38265 (ingress) -> test-zone/httpbin-5668f47445-mk7r2:80 (ID:96395) to-endpoint FORWARDED (TCP Flags: SYN)
Feb  5 21:17:44.348: 10.42.1.45:38265 (ingress) <- test-zone/httpbin-5668f47445-mk7r2:80 (ID:96395) to-network FORWARDED (TCP Flags: SYN, ACK)
Feb  5 21:17:49.349: 192.168.1.177:65017 (ingress) -> cilium-gateway/cilium-gateway-internal:443 (world) http-request FORWARDED (HTTP/1.1 GET https://httpbin.mydomain.tld/)
Feb  5 21:17:49.350: 192.168.1.177:65017 (ingress) <- cilium-gateway/cilium-gateway-internal:443 (world) http-response FORWARDED (HTTP/1.1 503 5001ms (GET https://httpbin.mydomain.tld/))
Feb  5 21:17:49.390: 10.42.1.45:34395 (ingress) -> test-zone/httpbin-5668f47445-mk7r2:80 (ID:96395) to-endpoint FORWARDED (TCP Flags: SYN)
Feb  5 21:17:49.390: 10.42.1.45:34395 (ingress) <- test-zone/httpbin-5668f47445-mk7r2:80 (ID:96395) to-network FORWARDED (TCP Flags: SYN, ACK)
Feb  5 21:17:51.642: 10.42.1.45:38265 (ingress) <- test-zone/httpbin-5668f47445-mk7r2:80 (ID:96395) to-network FORWARDED (TCP Flags: SYN, ACK)
Feb  5 21:17:56.546: 10.42.1.45:34395 (ingress) <- test-zone/httpbin-5668f47445-mk7r2:80 (ID:96395) to-network FORWARDED (TCP Flags: SYN, ACK)
Feb  5 21:17:59.745: 10.42.1.45:38265 (ingress) <- test-zone/httpbin-5668f47445-mk7r2:80 (ID:96395) to-network FORWARDED (TCP Flags: SYN, ACK)
Feb  5 21:18:04.652: 10.42.1.45:34395 (ingress) <- test-zone/httpbin-5668f47445-mk7r2:80 (ID:96395) to-network FORWARDED (TCP Flags: SYN, ACK)
Feb  5 21:18:16.385: 10.42.1.45:38265 (ingress) <- test-zone/httpbin-5668f47445-mk7r2:80 (ID:96395) to-network FORWARDED (TCP Flags: SYN, ACK)
```

Functional on every other node
```log
hubble observe -f --identity 8
Feb  5 17:28:25.419: 10.42.1.45:39087 (ingress) -> test-zone/echo-679b5c479f-ct8rr:80 (ID:90104) to-endpoint FORWARDED (TCP Flags: SYN)
Feb  5 17:28:25.419: 10.42.1.45:39087 (ingress) <- test-zone/echo-679b5c479f-ct8rr:80 (ID:90104) to-network FORWARDED (TCP Flags: SYN, ACK)
Feb  5 17:28:25.419: 10.42.1.45:39087 (ingress) -> test-zone/echo-679b5c479f-ct8rr:80 (ID:90104) to-endpoint FORWARDED (TCP Flags: ACK)
Feb  5 17:28:25.420: 10.42.1.45:39087 (ingress) -> test-zone/echo-679b5c479f-ct8rr:80 (ID:90104) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Feb  5 17:28:25.423: 10.42.1.45:39087 (ingress) <> test-zone/echo-679b5c479f-ct8rr (ID:90104) pre-xlate-rev TRACED (TCP)
Feb  5 17:28:25.427: 10.42.1.45:39087 (ingress) <- test-zone/echo-679b5c479f-ct8rr:80 (ID:90104) to-network FORWARDED (TCP Flags: ACK, PSH)
Feb  5 17:28:30.433: 10.42.1.45:39087 (ingress) <- test-zone/echo-679b5c479f-ct8rr:80 (ID:90104) to-network FORWARDED (TCP Flags: ACK, FIN)
Feb  5 17:28:30.434: 10.42.1.45:39087 (ingress) -> test-zone/echo-679b5c479f-ct8rr:80 (ID:90104) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
```

I have tested with Cilium 1.15.0 on Kubernetes 1.29.0-k3s1.  Nodes using linux kernel 5.10.160-rockchip
```yaml
autoDirectNodeRoutes: true
bandwidthManager:
  enabled: true
  bbr: false
bpf:
  masquerade: true
cluster:
  name: home-cluster
  id: 1
containerRuntime:
  integration: containerd
  socketPath: /var/run/k3s/containerd/containerd.sock
endpointRoutes:
  enabled: true
hubble:
  enabled: true
  metrics:
    enabled:
      - dns:query
      - drop
      - tcp
      - flow
      - port-distribution
      - icmp
      - http
  relay:
    enabled: true
    rollOutPods: true
  ui:
    enabled: true
    rollOutPods: true
    ingress:
      enabled: false
ipam:
  mode: kubernetes
ipv4NativeRoutingCIDR: "10.42.0.0/16"
k8sServiceHost: "192.168.1.195"
k8sServicePort: 6443
kubeProxyReplacement: true
kubeProxyReplacementHealthzBindAddr: 0.0.0.0:10256
l2announcements:
  enabled: true
loadBalancer:
  algorithm: maglev
  mode: dsr
localRedirectPolicy: true
operator:
  replicas: 1
  rollOutPods: true
rollOutCiliumPods: true
securityContext:
  privileged: true
routingMode: native
gatewayAPI:
  enabled: true
  secretsNamespace:
    create: false
    name: cilium-secrets
ingressController:
  enabled: false
  loadbalancerMode: shared
  enforceHttps: false
  secretsNamespace:
    create: false
    name: cilium-secrets
  service:
    allocateLoadBalancerNodePorts: false
```

and with some of the most recent images from the main branch ci https://github.com/cilium/cilium/actions/runs/7788397285/job/21237711996 using the following values
```yaml
image:
  override: "quay.io/cilium/cilium-ci:840cc579b7b5aac24ba00c4d8c8f1d10334882fa"
operator:
  image:
    override: "quay.io/cilium/operator-generic-ci:840cc579b7b5aac24ba00c4d8c8f1d10334882fa"
 ```
 
 
 
 So far the only workaround I have found is keep restarting cilium daemonset pods until the 'bad' pod is on a Kubernetes control-plane that hosts no website application pods, mitigating the impact.

**Comment by maxpain** [comment]
Any updates?

**Comment by larssb** [comment]
I was hoping that https://github.com/cilium/cilium/pull/32367 released in v1.15.6 would solve this. Didn't. So what we do to currently solve it is also to use hostLegacyRouting. This is annoying as that is less performant. But, we can live with it on the clusters whereon we use the Gateway API through Cilium.

**Comment by joelMuehlena** [comment]
I also ran in this Kind of issue today using `hostLegacyRouting: false` and IPv6. Unfortunately GatewayAPI is not possible to be used this way in some cases because it is very unreliable.

Edit: Using plain LoadBalancer services or Port-Forwarding works, so the issue most likely is rootet in envoy which returns `upstream connect error or disconnect/reset before headers. reset reason: remote connection failure, transport failure reason: delayed connect error: Connection refused`

```
cilium-envoy-bjg54 cilium-envoy 2025-09-19T14:46:54.722523145+02:00 [2025-09-19 12:46:54.722][22][trace][filter] [cilium/bpf_metadata.cc:537] onAccept (socket=73)
cilium-envoy-bjg54 cilium-envoy 2025-09-19T14:46:54.722529699+02:00 [2025-09-19 12:46:54.722][22][debug][config] [cilium/bpf_metadata.cc:400] EGRESS POD IP: 2001:470:735e:4:fde9:f673:7498:5171, destination IP: 2001:470:735e:ffff::1 sni: "pdns-auth.services.homelab.muehlena.de"
cilium-envoy-bjg54 cilium-envoy 2025-09-19T14:46:54.722533823+02:00 [2025-09-19 12:46:54.722][22][debug][filter] [cilium/conntrack.cc:189] cilium.bpf_metadata: Using conntrack map global
cilium-envoy-bjg54 cilium-envoy 2025-09-19T14:46:54.722687715+02:00 [2025-09-19 12:46:54.722][22][debug][filter] [cilium/conntrack.cc:158] cilium.bpf_metadata: Opened IPv6 conntrack map global
cilium-envoy-bjg54 cilium-envoy 2025-09-19T14:46:54.722701833+02:00 [2025-09-19 12:46:54.722][22][debug][filter] [cilium/conntrack.cc:258] cilium.bpf_metadata: IPv6 conntrack map global lookup failed: No such file or directory
```

Edit2: Kept debugging and found out: Envoy sometimes proxies the request with IPv6 sometimes with IPv6. The connection refused indeed came from the service, but sometimes it worked because the request happened with IPv4 instead IPv6 (looks like every 2nd request switchted IP protocols). So not related to this here.
